tcmur_cmd_handler.c

```diff
@@ -181,6 +181,340 @@ struct write_same {
+	 * equals to UNMAP GRANULARITY ALIGNMENT to simplify the
+	 * calculate algorithm, but in future for some new devices
+	 * who could support and they must have different values
+	 * to make the unmap to be more efficient, the following
+	 * align and split calculate algorithm should be changed.
+	 */
+	lbas = opt_unmap_gran - (lba & mask);
+	lbas = min(lbas, nlbas);
+
+	while (nlbas) {
+		desc = calloc(1, sizeof(*desc));
+		if (!desc) {
+			tcmu_dev_err(dev, "Failed to calloc desc!\n");
+			return tcmu_set_sense_data(sense, HARDWARE_ERROR,
+						   ASC_INTERNAL_TARGET_FAILURE,
+						   NULL);
+		}
+
+		ucmd = calloc(1, sizeof(*ucmd));
+		if (!ucmd) {
+			tcmu_dev_err(dev, "Failed to calloc unmapcmd!\n");
+			ret = tcmu_set_sense_data(sense, HARDWARE_ERROR,
+						  ASC_INTERNAL_TARGET_FAILURE,
+						  NULL);
+			goto free_desc;
+		}
+
+		desc->origcmd = origcmd;
+		desc->offset = lba * block_size;
+		desc->length = lbas * block_size;
+		ucmd->cmdstate = desc;
+
+		tcmu_dev_dbg(dev, "Split %d: start lba: %llu, end lba: %llu, lbas: %u\n",
+			     j++, lba, lba + lbas - 1, lbas);
+
+		ret = async_handle_cmd(dev, ucmd, unmap_work_fn);
+		if (ret != TCMU_ASYNC_HANDLED) {
+			state->error = true;
+			goto free_ucmd;
+		}
+
+		nlbas -= lbas;
+		lba += lbas;
+
+		lbas = min(opt_unmap_gran, nlbas);
+
+		state->refcount++;
+	}
+
+	return ret;
+
+free_ucmd:
+	free(ucmd);
+free_desc:
+	free(desc);
+	return ret;
+}
+
+static int handle_unmap_internal(struct tcmu_device *dev, struct tcmulib_cmd *origcmd,
+				 uint16_t bddl, uint8_t *par)
+{
+	struct unmap_state *state = origcmd->cmdstate;
+	uint8_t *sense = origcmd->sense_buf;
+	uint16_t offset = 0;
+	int ret, i = 0, refcount;
+
+	/* The first descriptor list offset is 8 in Data-Out buffer */
+	par += 8;
+
+	pthread_mutex_lock(&state->lock);
+	while (bddl) {
+		uint64_t lba;
+		uint64_t nlbas;
+
+		lba = be64toh(*((uint64_t *)&par[offset]));
+		nlbas = be32toh(*((uint32_t *)&par[offset + 8]));
+
+		tcmu_dev_dbg(dev, "Parameter list %d, start lba: %llu, end lba: %llu, nlbas: %u\n",
+			     i++, lba, lba + nlbas - 1, nlbas);
+
+		if (nlbas > VPD_MAX_UNMAP_LBA_COUNT) {
+			tcmu_dev_err(dev, "Illegal parameter list LBA count %lu exceeds:%u\n",
+				     nlbas, VPD_MAX_UNMAP_LBA_COUNT);
+			ret = tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+						  ASC_INVALID_FIELD_IN_PARAMETER_LIST,
+						  NULL);
+			goto state_unlock;
+		}
+
+		ret = check_lbas(dev, lba, nlbas);
+		if (ret) {
+			ret = tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+						  ASC_LBA_OUT_OF_RANGE, NULL);
+			goto state_unlock;
+		}
+
+		ret = handle_split_align_unmap(dev, origcmd, lba, nlbas);
+		if (ret != TCMU_ASYNC_HANDLED)
+			goto state_unlock;
+
+		/* The unmap block descriptor data length is 16 */
+		offset += 16;
+		bddl -= 16;
+	}
+state_unlock:
+	refcount = state->refcount;
+	pthread_mutex_unlock(&state->lock);
+
+	/* Or will let the cbk to do the release */
+	if (!refcount)
+		free(state);
+
```

Comment:
I think this still has the bug fixed in my patch

https://github.com/open-iscsi/tcmu-runner/pull/243

where if we failed after we have done a couple unmap calls, then we will return failure here, track_aio_request_finish will be called and the this code patch will complete the command. But refcount will be non-zero and so handle_unmap_cbk will also call the aio track function and complete the command.